### PR TITLE
drivers: retained_mem: ram: add support for nRF54L15

### PR DESCRIPTION
**Pull request by nika-nordic**
Add support for retained memory driver using nrfx_ram_ctrl helper.

### CODE REVIEW

File: samples/boards/nrf/system_off/src/retained_mem.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/crc.h>
+
+const static struct device *retained_mem_device = DEVICE_DT_GET(DT_ALIAS(retainedmemdevice));
+
+struct retained_data retained;
```
Comment (hubertmis) [inline samples/boards/nrf/system_off/src/retained_mem.c:20]: Is there a reason for this variable to be in the global scope?

**Review by hubertmis — DISMISSED** [review]

File: samples/boards/nrf/system_off/src/retained_mem.c (hunk 1 above)
Comment (nika-nordic) [reply samples/boards/nrf/system_off/src/retained_mem.c:20]: yes - this structure is referenced directly by sample `main.c` code: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/nrf/system_off/src/main.c#L40-L43 and this is a scheme enforced by existing POWER-based implementation for nRF52: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/boards/nrf/system_off/src/retained.c

I would like to avoid major refactors of this sample, as the goal was to add support for nRF54L15 rather than making it "perfect"

File: samples/boards/nrf/system_off/src/retained_mem.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+#define RETAINED_CRC_OFFSET offsetof(struct retained_data, crc)
+#define RETAINED_CHECKED_SIZE (RETAINED_CRC_OFFSET + sizeof(retained.crc))
+
+bool retained_validate(void)
```
Comment (nordic-krch) [inline samples/boards/nrf/system_off/src/retained_mem.c:25]: this function looks very like https://github.com/zephyrproject-rtos/zephyr/blob/34e4edc28a3806885821d863cc3ebc9c60f5b817/samples/boards/nrf/system_off/src/retained.c#L139 Why not merge this into a single file and avoid redundancy?

File: tests/drivers/retained_mem/api/testcase.yaml [hunk 3]
```diff
@@ -17,3 +17,10 @@ tests:
     tags:
       - drivers
       - retained_mem
+  drivers.retained_mem.api.ram.nrf54l15pdk_nrf54l15_cpuapp:
+    platform_allow: nrf54l15pdk_nrf54l15_cpuapp
+    extra_args:
+      - DTC_OVERLAY_FILE="nrf54l15pdk_nrf54l15_cpuapp_ram.overlay"
```
Comment (nordic-krch) [inline tests/drivers/retained_mem/api/testcase.yaml:23]: why not remove `_ram` suffix, remove this  configuration and just add `nrf54l15pdk_nrf54l15_cpuapp` to `platform_allow` for `drivers.retained_mem.api.ram` case.

File: drivers/retained_mem/Kconfig.nrf [hunk 4]
```diff
@@ -8,3 +8,10 @@ config RETAINED_MEM_NRF_GPREGRET
 	help
 	  Enable driver for Nordic nRF GPREGRET-based retained memory
 	  register support.
+
+config RETAINED_MEM_NRF_RAM_CTRL
+	bool "nRF retention driver based on RAM control helper"
```
Comment (nordic-krch) [inline drivers/retained_mem/Kconfig.nrf:13]: imo this is to detailed description. Maybe just `nRF RAM retention driver`? User does not know and does not care what is `RAM control helper`.

File: samples/boards/nrf/system_off/boards/nrf54l15pdk_nrf54l15_cpuapp_retained_mem.overlay [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+	aliases {
+		retainedmemdevice = &retainedmem0;
+	};
+};
+
+&cpuapp_sram {
+	reg = <0x20000000 DT_SIZE_K(184)>;
+	ranges = <0x0 0x20000000 0x2e000>;
+};
```
Comment (masz-nordic) [inline samples/boards/nrf/system_off/boards/nrf54l15pdk_nrf54l15_cpuapp_retained_mem.overlay:22]: It would be useful to add a comment why this is needed.

File: samples/boards/nrf/system_off/src/retained_mem.c (hunk 2 above)
Comment (masz-nordic) [reply samples/boards/nrf/system_off/src/retained_mem.c:25]: I agree, the difference between those two is so small they should be merged.

File: drivers/retained_mem/retained_mem_nrf_ram_ctrl.c [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+	for (size_t i = 0; i < ARRAY_SIZE(ret_mem_regions); i++) {
+		rmr = &ret_mem_regions[i];
+		nrfx_ram_ctrl_retention_enable_set((void *)rmr->dt_addr, rmr->dt_size, true);
+	}
+
+	return 0;
+}
+
+SYS_INIT(retained_mem_nrf_init, POST_KERNEL, 0);
```
Comment (nordicjm) [inline drivers/retained_mem/retained_mem_nrf_ram_ctrl.c:37]: I would think this should be configured earlier

File: samples/boards/nrf/system_off/src/retained_mem.c (hunk 2 above)
Comment (nika-nordic) [reply samples/boards/nrf/system_off/src/retained_mem.c:27]: fixed

File: tests/drivers/retained_mem/api/testcase.yaml (hunk 3 above)
Comment (nika-nordic) [reply tests/drivers/retained_mem/api/testcase.yaml]: fixed

File: drivers/retained_mem/Kconfig.nrf (hunk 4 above)
Comment (nika-nordic) [reply drivers/retained_mem/Kconfig.nrf]: fixed

File: drivers/retained_mem/retained_mem_nrf_ram_ctrl.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (nika-nordic) [reply drivers/retained_mem/retained_mem_nrf_ram_ctrl.c]: fixed

File: samples/boards/nrf/system_off/boards/nrf54l15pdk_nrf54l15_cpuapp_retained_mem.overlay (hunk 5 above)
Comment (nika-nordic) [reply samples/boards/nrf/system_off/boards/nrf54l15pdk_nrf54l15_cpuapp_retained_mem.overlay]: fixed

File: drivers/retained_mem/retained_mem_nrf_ram_ctrl.c (hunk 6 above)
Comment (nika-nordic) [reply drivers/retained_mem/retained_mem_nrf_ram_ctrl.c]: fixed

File: samples/boards/nrf/system_off/README.rst [hunk 8]
```diff
@@ -12,16 +12,16 @@ deep sleep on Nordic platforms.
 RAM Retention
 =============
 
-On nRF52 platforms this also can demonstrate RAM retention.  By selecting
-``CONFIG_APP_RETENTION=y`` state related to number of boots, number of times
-system off was entered, and total uptime since initial power-on are retained
-in a checksummed data structure.  The POWER peripheral is configured to keep
-the containing RAM section powered while in system-off mode.
+This sample can also can demonstrate RAM retention. By selecting
+``CONFIG_APP_USE_NRF_RETENTION=y`` or ``CONFIG_APP_USE_RETAINED_MEM``
```
[the remark this answers is not in the document]
Comment (nika-nordic) [reply samples/boards/nrf/system_off/README.rst]: fixed

**Review by nordic-krch — DISMISSED** [review]

**Review by masz-nordic — DISMISSED** [review]

**Review by nordic-krch — DISMISSED** [review]

File: drivers/retained_mem/Kconfig.nrf [hunk 9]
```diff
@@ -8,3 +8,10 @@ config RETAINED_MEM_NRF_GPREGRET
 	help
 	  Enable driver for Nordic nRF GPREGRET-based retained memory
 	  register support.
+
+config RETAINED_MEM_NRF_RAM_CTRL
+	bool "nRF RAM retention driver"
+	default y
+	depends on DT_HAS_ZEPHYR_RETAINED_RAM_ENABLED
+	help
```
Comment (nordicjm) [inline drivers/retained_mem/Kconfig.nrf:16]: this needs `depends on SOC_NRF54L15` or `depends on SOC_SERIES_NRF54LX` (whichever if applicable)

Comment (nordicjm) [reply drivers/retained_mem/Kconfig.nrf:16]: Also needs to depend on `RETAINED_MEM_ZEPHYR_RAM`

Comment (nika-nordic) [reply drivers/retained_mem/Kconfig.nrf:18]: done

File: drivers/retained_mem/Kconfig.nrf [hunk 10]
```diff
@@ -8,3 +8,11 @@ config RETAINED_MEM_NRF_GPREGRET
 	help
 	  Enable driver for Nordic nRF GPREGRET-based retained memory
 	  register support.
+
+config RETAINED_MEM_NRF_RAM_CTRL
+	bool "nRF RAM retention driver"
+	default y
+	depends on DT_HAS_ZEPHYR_RETAINED_RAM_ENABLED && RETAINED_MEM_ZEPHYR_RAM
+	depends on SOC_SERIES_NRF52X || SOC_SERIES_NRF53X || SOC_SERIES_NRF54LX || SOC_SERIES_NRF91X
```
Comment (nordicjm) [inline drivers/retained_mem/Kconfig.nrf:16]: `depends on SOC_SERIES_NRF52X || SOC_SERIES_NRF53X || SOC_SERIES_NRF54LX || SOC_SERIES_NRF91X`

What? Why is this needed on anything that isn't nrf54l?

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: drivers/retained_mem/Kconfig.nrf (hunk 10 above)
Comment (nika-nordic) [reply drivers/retained_mem/Kconfig.nrf:16]: Purpose of this PR was to mainly support nrf54l, however the updated `retained_mem` driver supports nrf52/nrf53/nrf91 as well, hence this list of ORed SoC series

Comment (nordicjm) [reply drivers/retained_mem/Kconfig.nrf:16]: Support how? The RAM is retained by default which is what I assume `nrfx_ram_ctrl_retention_enable_set` sets, so why is that needed when it's already enabled?

Comment (nordicjm) [reply drivers/retained_mem/Kconfig.nrf:16]: All RAM is enabled by default, I don't mean retention I just mean RAM itself, which means you can reboot an application and keep the contents, this definitely works because we have 2 tests that check it on real hardware in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/boot/mcuboot_data_sharing and https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/boot/mcuboot_recovery_retention

So this allows retaining RAM contents if someone turns the RAM section off... but why? If someone turns RAM sections off, they should know what they are doing, and having a part that is retained that you can't do anything with makes no sense to me. Not to mention that as I found out and posted on NCSDK-22549 after a gruelling battle with an ncs sample application and debugger, turning off RAM sections is a great way to lock the whole CPU up - or brick the module.

Comment (hubertmis) [reply drivers/retained_mem/Kconfig.nrf:16]: This is about preserving partial RAM content when using https://docs.zephyrproject.org/apidoc/latest/group__sys__poweroff.html#gaf4552a8ea40b4adb94f85b0a1ff06e39

This function results in SystemOff state in nRF52/53/91 in which the RAM is powered down and its content is unretained by default.

Comment (nordicjm) [reply drivers/retained_mem/Kconfig.nrf:16]: Interesting, then this should depends on `POWEROFF`

Comment (nika-nordic) [reply drivers/retained_mem/Kconfig.nrf:16]: added dependency on `POWEROFF`

**Review by nordic-krch — DISMISSED** [review]

File: drivers/retained_mem/Kconfig.nrf (hunk 10 above)
Comment (nordicjm) [reply drivers/retained_mem/Kconfig.nrf:16]: you have added the wrong dependency, it needs to be `POWEROFF` not `HAS_POWEROFF`

**Review by nordicjm — CHANGES_REQUESTED** [review]

**Review by hubertmis — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]

**Review by kl-cruz — APPROVED** [review]
